docs/compatibility.md

```diff
@@ -0,0 +1,185 @@
+
+\*StableHLO semantics is defined by the StableHLO specification and can be tested using the StableHLO interpreter. While we don't have a specification, we will be referring to a combination of [XLA's Operation Semantics](https://www.tensorflow.org/xla/operation_semantics) and existing implementations to define the semantics.
+
+## StableHLO Programs
+
+For the purposes of compatibility guarantees, we define StableHLO programs as MLIR programs that include ops, attributes, and types specified below. This definition is based on our experience with lowering JAX/PyTorch/TensorFlow programs to MHLO and interviewing various groups using MHLO. Please let us know if you're interested in expanding this definition, and we can have a discussion about this.
+
+|  Dialect     | Ops, attributes and types |
+| ------------ | ----------- |
+| Arith        | `AddIOp, CmpIOp, CmpIPredicate, ConstantOp, DivSIOp, ExtSIOp, IndexCastOp, MaxSIOp, MinSIOp, MulIOp, SelectOp, SubIOp, TruncIOp` |
+| Builtin      | No ops, but all attributes and types. |
+| CHLO         | All ops, attributes and types. |
+| Func         | `CallOp` |
+| MLProgram    | All ops, attributes and types. |
+| SparseTensor | Aspirational (pending the sparsity RFC which is expected in Q4 2022). |
+| StableHLO    | All ops, attributes and types except `CustomCallOp` whose semantics is implementation-defined. |
+| Tensor       | `CastOp, DimOp, FromElementsOp` |
+
+We provide a pass `--stablehlo-compatibility-check` which succeeds if a given program is fully covered by these compatibility guarantees and fails if it doesn't.
+
+## Compatibility Protocols
+
+### Compatible Changes
+_Characteristics of a compatible change:_
+
+- Does not break forward or backward compatibility.
+- May require updating Python/C APIs.
+
+_Example:_ Rename an operation, but not the mnemonic `SinOp → SineOp`:
+
+_Compatibility Protocol:_
+
+![Compatible Change Protocol Diagram](images/compatible_change.png)
+
+_Description:_
+- 1/1/23: Create bytecode_v1 from libStablehlo@v1
+- 1/7/23: Rename MyOp → TheOp, mnemonic remains MyDialect.MyOp in libStablehlo@v2
+  + No change to serialization / deserialization
+  + Requires changes to users of other libStablehlo APIs
+  + Backward compatibility: bytecode_v1 has the same semantics in libStablehlo@v2.
+  + Forward compatibility: bytecode_v2 has the same semantics in libStablehlo@v1.
+
+### Migration Change
+_Characteristics of a migration change:_
+- Change in dialect breaks serialization in a semantically compatible manner (IR able to be upgraded/downgraded).
+- Downgrade bytecode for 3 weeks.
+- Upgrade bytecode for 6 months.
+
+_Example:_ 
+- Rename an Operation mnemonic, `cross-replica-sum → cross_replica_sum`.
+
+_Compatibility Protocol:_
+
+![Migration Change Protocol Diagram](images/migration_change.png)
+
+_Description:_
+- 1/1/23: Create bytecode_v1 from libStablehlo@v1
+- 1/7/2023: Rename op and mnemonic `OldOp → NewOp`, build libStablehlo@v2
```

Comment:
Given that we're using the OldOp / NewOp wording here, perhaps let's also use it in the first and second pictures? I think this will help with comprehension, since it's not evident which of the MyOp / TheOp ops is old and which is new.

Reply:
updated images